mobile/lib/modules/home/ui/asset_grid/asset_grid_data_structure.dart

```diff
@@ -2,212 +2,282 @@ import 'dart:math';
   final String? title;
   final DateTime date;
-  final List<Asset>? relatedAssetList;
+  final int count;
+  final int offset;
+  final int totalCount;
 
   RenderAssetGridElement(
     this.type, {
-    this.assetRow,
     this.title,
     required this.date,
-    this.relatedAssetList,
+    this.count = 0,
+    this.offset = 0,
+    this.totalCount = 0,
   });
 }
 
 enum GroupAssetsBy {
   day,
-  month;
-}
-
-class AssetGridLayoutParameters {
-  final int perRow;
-  final bool dynamicLayout;
-  final GroupAssetsBy groupBy;
-
-  AssetGridLayoutParameters(
-    this.perRow,
-    this.dynamicLayout,
-    this.groupBy,
-  );
-}
-
-class _AssetGroupsToRenderListComputeParameters {
-  final List<Asset> assets;
-  final AssetGridLayoutParameters layout;
-
-  _AssetGroupsToRenderListComputeParameters(
-    this.assets,
-    this.layout,
-  );
+  month,
+  auto,
+  none,
+  ;
 }
 
 class RenderList {
   final List<RenderAssetGridElement> elements;
-
-  RenderList(this.elements);
-
-  static Map<DateTime, List<Asset>> _groupAssets(
-    List<Asset> assets,
-    GroupAssetsBy groupBy,
-  ) {
-    if (groupBy == GroupAssetsBy.day) {
-      return assets.groupListsBy(
-        (element) {
-          final date = element.fileCreatedAt.toLocal();
-          return DateTime(date.year, date.month, date.day);
-        },
-      );
-    } else if (groupBy == GroupAssetsBy.month) {
-      return assets.groupListsBy(
-        (element) {
-          final date = element.fileCreatedAt.toLocal();
-          return DateTime(date.year, date.month);
-        },
-      );
+  final List<Asset>? allAssets;
+  final QueryBuilder<Asset, Asset, QAfterSortBy>? query;
+  final int totalAssets;
+  List<Asset> _buf = [];
+  int _bufOffset = 0;
+
+  RenderList(this.elements, this.query, this.allAssets)
+      : totalAssets = allAssets?.length ?? query!.countSync();
+
+  bool get isEmpty => totalAssets == 0;
+
+  List<Asset> loadAssets(int offset, int count) {
+    assert(offset >= 0);
+    assert(count > 0);
+    assert(offset + count <= totalAssets);
+    if (allAssets != null) {
+      return allAssets!.slice(offset, offset + count);
+    } else if (query != null) {
+      if (offset < _bufOffset || offset + count > _bufOffset + _buf.length) {
+        final bool forward = _bufOffset < offset;
+        const batchSize = 128;
+        const oppositeSize = 32;
+        final len = max(batchSize, count + oppositeSize);
+        final start = max(
+          0,
+          forward
+              ? offset - oppositeSize
+              : (len > batchSize ? offset : offset + count - len),
+        );
+        _buf = query!.offset(start).limit(len).findAllSync();
+        _bufOffset = start;
+      }
+      assert(_bufOffset <= offset);
+      assert(_bufOffset + _buf.length >= offset + count);
+      final arr = _buf.slice(offset - _bufOffset, offset - _bufOffset + count);
+      return arr;
     }
+    throw Exception("RenderList has neither assets nor query");
+  }
```

Comment:
Can you add some comments explaining your algorithm, please? This code is (at least for me) very complex and hard to understand.

Reply:
sure, added comments